x-pack/plugins/lens/public/state_management/load_initial.test.tsx

```diff
@@ -0,0 +1,323 @@
+    expect(deps.datasourceMap.testDatasource.initialize).toHaveBeenCalled();
+    expect(deps.datasourceMap.testDatasource2.initialize).not.toHaveBeenCalled();
+    expect(deps.visualizationMap.testVis.initialize).toHaveBeenCalled();
+    expect(deps.visualizationMap.testVis2.initialize).not.toHaveBeenCalled();
+  });
+
+  it('should initialize all datasources with state from doc', async () => {
+    const datasource1State = { datasource1: '' };
+    const datasource2State = { datasource2: '' };
+    const services = makeDefaultServices();
+    services.attributeService.unwrapAttributes = jest.fn().mockResolvedValue({
+      exactMatchDoc,
+      visualizationType: 'testVis',
+      title: '',
+      state: {
+        datasourceStates: {
+          testDatasource: datasource1State,
+          testDatasource2: datasource2State,
+        },
+        visualization: {},
+        query: { query: '', language: 'lucene' },
+        filters: [],
+      },
+      references: [],
+    });
+
+    const storeDeps = mockStoreDeps({
+      lensServices: services,
+      visualizationMap: {
+        testVis: {
+          ...createMockVisualization(),
+          id: 'testVis',
+          visualizationTypes: [
+            {
+              icon: 'empty',
+              id: 'testVis',
+              label: 'TEST1',
+              groupLabel: 'testVisGroup',
+            },
+          ],
+        },
+      },
+      datasourceMap: {
+        testDatasource: createMockDatasource('testDatasource'),
+        testDatasource2: createMockDatasource('testDatasource2'),
+        testDatasource3: createMockDatasource('testDatasource3'),
+      },
+    });
+
+    const { store, deps } = await makeLensStore({
+      storeDeps,
+      preloadedState,
+    });
+
+    await act(async () => {
+      await store.dispatch(loadInitial(defaultProps));
+    });
+    const { datasourceMap } = deps;
+    expect(datasourceMap.testDatasource.initialize).toHaveBeenCalled();
+
+    expect(datasourceMap.testDatasource.initialize).toHaveBeenCalledWith(
+      datasource1State,
+      [],
+      undefined,
+      {
+        isFullEditor: true,
+      }
+    );
+    expect(datasourceMap.testDatasource2.initialize).toHaveBeenCalledWith(
+      datasource2State,
+      [],
+      undefined,
+      {
+        isFullEditor: true,
+      }
+    );
+    expect(datasourceMap.testDatasource3.initialize).not.toHaveBeenCalled();
+    expect(store.getState()).toMatchSnapshot();
+  });
+
+  describe('loadInitial', () => {
+    it('does not load a document if there is no initial input', async () => {
+      const { deps, store } = makeLensStore({ preloadedState });
+      await act(async () => {
+        await store.dispatch(
+          loadInitial({
+            ...defaultProps,
+            initialInput: undefined,
+          })
+        );
+      });
+      expect(deps.lensServices.attributeService.unwrapAttributes).not.toHaveBeenCalled();
+    });
+
+    it('starts new searchSessionId', async () => {
```

Comment:
this is the new test that tests the bug this PR fixes. The rest is refactored to test the whole path from invoking loadInitial (we do it in mounter) to run through middleware and reducer (before it was only checking middleware in separation)